docs/lang/articles/advanced/sparse.md

```diff
@@ -2,100 +2,116 @@
-compared to quadtrees and octrees, it is more performance-friendly to use shallower trees with larger branching factors.
-[VDB](https://www.openvdb.org/) and [SPGrid](http://pages.cs.wisc.edu/~sifakis/papers/SPGrid.pdf) are such examples.
-In Taichi, programmers can compose data structures similar to VDB and SPGrid with SNodes.
+:::note
+Reading an inactive voxel returns zero.
+:::
 
-![image](https://raw.githubusercontent.com/taichi-dev/public_files/master/taichi/doc/sparse_grids_3d.jpg)
-Figure: A 3D fluid simulation that uses both particles and grids. Left to right: particles, 1x1x1 voxels, 4x4x4 blocks, 16x16x16 blocks.
+### Pointer SNode
 
-#### Blocked leaf cells and bitmasks
+```python {2} title=pointer.py
+x = ti.field(ti.f32)
+block = ti.root.pointer(ti.ij, (4,4))
+pixel = block.dense(ti.ij, (2,2))
+pixel.place(x)
+
+@ti.kernel
+def activate():
+    x[2,3] = 1.0
+    x[2,4] = 2.0
+
+@ti.kernel
+def print_active():
+    for i, j in block:
+        print("Active block", i, j)
+    # output: Active block 1 1
+    #         Active block 1 2
+    for i, j in x:
+        print('field x[{}, {}] = {}'.format(i, j, x[i, j]))
+    # output: field x[2, 2] = 0.000000
+    #         field x[2, 3] = 1.000000
+    #         field x[3, 2] = 0.000000
+    #         field x[3, 3] = 0.000000
+    #         field x[2, 4] = 2.000000
+    #         field x[2, 5] = 0.000000
+    #         field x[3, 4] = 0.000000
+    #         field x[3, 5] = 0.000000
+```
+The code snippet above creates an 8x8 sparse grid, with the top-level being 4x4 pointer arrays (line 2 of `pointer.py`),
+and each pointer points to a 2x2 dense block. The `activate` function could automatically activates `block[1,1]` which includes `x[2,3]` and `block[1,2]` which includes `x[2, 4]`. Other pixels of `block[1,1]` (`x[2,2], x[3, 2], x[3, 3]`) and `block[1,2]` (`x[2, 5], x[3, 4], x[3, 5]`) are also implicitly activated if `pixel` is a dense node. Because all pixels in the dense block share the same activity value.
+As the below figure shown, the active blocks and pixels are green. You can write and read the sparse field like a dense field using indices.
+
+<center>
+
+![Pointer](https://github.com/FantasyVR/public_files/blob/sparse_computation/taichi/doc/pointer_small.png)
+
+</center>
+
+
+In fact, the sparse field is a Snode tree shown in the following figure. You could use the struct-for loop to loop over the different levels of the SNode tree like the `print_active` function in `pointer.py`. `for i, j in block` would loop over all active `pointer` Snodes (green grids). `for i, j in pixel` would loop over all active `dense` Snodes (green grids).
+
+<center>
+
+![Pointer Snode Tree](https://github.com/FantasyVR/public_files/blob/sparse_computation/taichi/doc/pointer_tree_small.png)
+
+</center>
 
 While a null pointer can effectively represent an empty sub-tree, at the leaf level using 64 bits to represent the activity
```

Comment:
This part should be moved under `bitmasked SNode`.

Reply:
This figure is used to show the pointer SNode tree.

Reply:
Oh I mean paragraphs starting from L116. GitHub shows lines above it, which might be a little confusing :-)

Reply:
I see. Actually, this paragraph applies to both `pointer + dense` and `pointer + bitmasked` examples.

Reply:
Yes but the main purpose is to introduce `bitmasked`. So I suggest moving all of them under `bitmasked` instead of splitting them in two places.